cast integer to string in case date time format is „U“

PercyDwn opened the pull request:
In case the date time format is „U“, the data comes as a integer (timestamp in seconds). To keep the „data is string“-condition, the integer is casted to a string.

| Q             | A
| ------------- | ---
| Branch?       | 6.3<!-- see below -->
| Bug fix?      | yes
| New feature?  | no <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #50236 <!-- prefix each issue number with "Fix #", no need to create an issue if none exists, explain below instead -->
| License       | MIT
| Doc PR        |  <!-- required for new features -->
<!--
Replace this notice by a short README for your feature/bugfix.
This will help reviewers and should be a good start for the documentation.

Additionally (see https://symfony.com/releases):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too).
 - Features and deprecations must be submitted against the latest branch.
 - For new features, provide some code snippets to help understand usage.
 - Changelog entry should follow https://symfony.com/doc/current/contributing/code/conventions.html#writing-a-changelog-entry
 - Never break backward compatibility (see https://symfony.com/bc).
-->

Fixes DateTimeNormalizer::denormalize throwing a NotNormalizableValueException when DateTimeFormat is "U" and data is passed as an integer.
The Exception should not be thrown, as data being an integer is valid for a timestamp.
My solution is to check if $dateTImeFormat is "U" and $data is an integer. In that case the integer is casted to a string. 
This way, the condition that $data is a string does not need to be changed.

PercyDwn committed:
cast integer to string in case date time format is „U“

In case the date time format is „U“, the data comes as a integer (timestamp in seconds).
To keep the „data is string“-condition, the integer is casted to a string.

stof commented:
I'd rather change the type check to also allow integers when using `U` instead of casting the value before the type check.

stof commented:
And this needs to have tests covering it to prevent regressions

PercyDwn commented:
Thank you for the feedback, @stof and @nicolas-grekas.

Is it the better solution to allow integers and floats for `U` and `U.u`? 
In the PHP documentation the datetime for `createFromFormat` is typed as a string:
```
public static DateTimeImmutable::createFromFormat(string $format, string $datetime, ?[DateTimeZone] $timezone = null): [DateTimeImmutable]|false
```
With PHP 8.1 following code works just fine (does not return false)
```
$object = DateTimeImmutable::createFromFormat("U", 1683635480);
$object = DateTimeImmutable::createFromFormat("U.u", 1683635480.111111);
```
but could this behavior change in the future?

stof commented:
@PercyDwn My guess is that this code would not work fine with strict types.

stof commented:
> Is it the better solution to allow integers and floats for `U` and `U.u`?

IMO, yes

PercyDwn committed:
allow string or integer with dateTimeFormat „U“ in type check

$data can come as an integer, when date time format is „U“. Updated type check, so in that case no exception is thrown.

PercyDwn committed:
fix coding standard

PercyDwn commented:
Okay so I updated the type check to allow integers when dateTimeFormat is "U". 
But I ignored "U.u" as writing a test for it was a total mess. I tried allowing floats with "U.u", but i have the feeling that in such a case the data would come as a string, not as a float. I was not able to write a working test where the time (including milliseconds) came as a float and was identical with the resulting denormalized date.

Some of the github tests are failing, but I am not quite sure why.

nicolas-grekas commented:
Fixed by #50251
